File: decider.py
"""Decision engine for the smart loop.

Two paths:
  - rule_decide: cheap, no model call. Returns a Decision when the answer
    is unambiguous (Recommended marker, cursor-on-Yes, canonical Yes/No).
  - llm_decide: calls a local Ollama text model with the parsed
    PromptFrame + project context.

The orchestrator (agent.py --smart) tries rule_decide first and only
falls through to llm_decide if the rule path returns None.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

import requests

from parser import PromptFrame, Option

OLLAMA_CHAT_URL = "http://localhost:11434/api/chat"
ALLOWED_ACTIONS = {"1", "2", "3", "4", "5", "y", "n", "a", "enter", "none"}
SYSTEM_PROMPT_PATH = Path(__file__).parent / "prompts" / "decide_system.txt"
DEFAULT_LLM_MODEL = "qwen3:latest"
DEFAULT_LLM_TIMEOUT_SEC = 30


@dataclass
class Decision:
    action: str        # one of ALLOWED_ACTIONS
    confidence: float  # 0..1
    reason: str
    source: str        # 'rule' | 'llm' | 'fallback'


SAFE_FIRST_LABEL = re.compile(r"^\s*(yes|allow|ok|continue|proceed)\b", re.IGNORECASE)
UNSAFE_FIRST_LABEL = re.compile(
    r"don'?t ask again|for the rest of|always allow|never ask",
    re.IGNORECASE,
)
# Strict canonical labels — matched as the entire stripped label.
CANONICAL_YES_LABEL = re.compile(
    r"^(yes|yes,?\s*proceed|yes,?\s*continue|allow|ok|continue|proceed)$",
    re.IGNORECASE,
)
CANONICAL_NO_LABEL_PREFIX = re.compile(r"^(no|cancel|abort)\b", re.IGNORECASE)


def _safe_default_option(frame: PromptFrame) -> Option | None:
    """Return the safest option to auto-pick, or None.

    Conservative — only fires when the assistant has clearly indicated the
    safe answer:
      (a) inline (y/n) prompt with literal 'yes' as the y option, OR
      (b) cursor sits on a Yes/Allow/OK/Continue/Proceed option, OR
      (c) exactly two options with canonical 'Yes' / 'No' labels and no
          'don't ask again' variants.

    All other ambiguous cases are deferred to the LLM path.
    """
    by_key = {o.key: o for o in frame.options}
    cand = by_key.get("1") or by_key.get("y")
    if not cand:
        return None
    if UNSAFE_FIRST_LABEL.search(cand.label):
        return None

    cand_label = cand.label.strip().lower()

    # (a) Inline (y/n) — synthesised by the parser with labels "yes"/"no".
    if (len(frame.options) == 2
            and {o.key for o in frame.options} == {"y", "n"}
            and cand.key == "y"
            and cand_label == "yes"):
        return cand

    # (b) Cursor on a Yes-style option.
    if cand.is_cursor and SAFE_FIRST_LABEL.search(cand.label):
        return cand

    # (c) Exactly two options, canonical Yes / No-prefixed labels.
    if (len(frame.options) == 2
            and cand.key == "1"
            and CANONICAL_YES_LABEL.match(cand_label)):
        other = by_key.get("2")
        if other and CANONICAL_NO_LABEL_PREFIX.match(other.label.strip()):
            return cand

    return None


def rule_decide(frame: PromptFrame) -> Decision | None:
    """Cheap, deterministic decisions. Returns None if LLM is needed."""
    if frame.kind != "choice":
        return Decision(
            action="none",
            confidence=1.0,
            reason=f"kind={frame.kind} → no auto-keypress",
            source="rule",
        )
    if frame.destructive:
        return Decision(
            action="none",
            confidence=1.0,
            reason="destructive keyword detected — handing to human",
            source="rule",
        )
    if not frame.options:
        return Decision(
            action="none", confidence=1.0,
            reason="no options parsed", source="rule",
        )

    recommended = [o for o in frame.options if o.recommended]
    if len(recommended) == 1:
        return Decision(
            action=recommended[0].key,
            confidence=0.95,
            reason=f"option {recommended[0].key!r} marked recommended",
            source="rule",
        )

    safe = _safe_default_option(frame)
    if safe:
        return Decision(
            action=safe.key,
            confidence=0.85,
            reason=f"safe single-shot Yes on key {safe.key!r}",
            source="rule",
        )
    return None


# ---------- LLM path ----------

def _format_frame_for_llm(frame: PromptFrame, project_context: str | None) -> str:
    parts = []
    if project_context:
        parts.append("PROJECT_CONTEXT:")
        parts.append(project_context.strip()[:1500])
        parts.append("")
    parts.append(f"PROMPT_QUESTION: {frame.question or '(none)'}")
    parts.append("PROMPT_OPTIONS:")
    for o in frame.options:
        marks = []
        if o.is_cursor:
            marks.append("cursor")
        if o.recommended:
            marks.append("recommended")
        suffix = f"  ({', '.join(marks)})" if marks else ""
        parts.append(f"  [{o.key}] {o.label}{suffix}")
    parts.append("")
    parts.append("RAW_OCR:")
    parts.extend(frame.raw_lines[-30:])  # cap to last 30 lines
    return "\n".join(parts)


def _load_system_prompt() -> str:
    return SYSTEM_PROMPT_PATH.read_text(encoding="utf-8")


def _call_ollama_chat(model: str, system: str, user: str,
                      timeout: int = DEFAULT_LLM_TIMEOUT_SEC) -> str:
    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": system},
            {"role": "user", "content": user},
        ],
        "stream": False,
        "format": "json",
        "options": {"temperature": 0.0},
    }
    r = requests.post(OLLAMA_CHAT_URL, json=payload, timeout=timeout)
    r.raise_for_status()
    return r.json().get("message", {}).get("content", "")


def _parse_llm_response(text: str) -> Decision | None:
    m = re.search(r"\{.*\}", text, re.DOTALL)
    if not m:
        return None
    try:
        obj = json.loads(m.group(0))
    except json.JSONDecodeError:
        return None
    action = str(obj.get("action", "")).strip().lower()
    if action not in ALLOWED_ACTIONS:
        return None
    try:
        conf = float(obj.get("confidence", 0.0))
    except (TypeError, ValueError):
        conf = 0.0
    reason = str(obj.get("reason", "")).strip()[:200]
    return Decision(action=action, confidence=conf, reason=reason, source="llm")


def llm_decide(
    frame: PromptFrame,
    project_context: str | None = None,
    *,
    model: str = DEFAULT_LLM_MODEL,
    confidence_threshold: float = 0.7,
    timeout: int = DEFAULT_LLM_TIMEOUT_SEC,
    chat_fn: Callable[[str, str, str], str] | None = None,
) -> Decision:
    """Call the local text LLM. `chat_fn` is injectable for tests."""
    if frame.kind != "choice":
        return Decision("none", 1.0, f"kind={frame.kind}", "fallback")

    system = _load_system_prompt()
    user = _format_frame_for_llm(frame, project_context)
    fn = chat_fn or (
        lambda s, u, _m=model, _t=timeout: _call_ollama_chat(_m, s, u, _t))
    try:
        raw = fn(system, user, model)
    except Exception as e:  # noqa: BLE001
        return Decision("none", 0.0, f"llm error: {e}", "fallback")

    parsed = _parse_llm_response(raw)
    if parsed is None:
        return Decision("none", 0.0, f"unparseable llm output: {raw[:120]!r}",
                        "fallback")
    if parsed.confidence < confidence_threshold:
        return Decision(
            "none", parsed.confidence,
            f"low confidence ({parsed.confidence:.2f}): {parsed.reason}",
            "llm",
        )
    return parsed


def decide(
    frame: PromptFrame,
    project_context: str | None = None,
    *,
    model: str = DEFAULT_LLM_MODEL,
    confidence_threshold: float = 0.7,
    timeout: int = DEFAULT_LLM_TIMEOUT_SEC,
    chat_fn: Callable[[str, str, str], str] | None = None,
) -> Decision:
    """Top-level: rule path → LLM path."""
    rule = rule_decide(frame)
    if rule is not None:
        return rule
    return llm_decide(
        frame, project_context,
        model=model,
        confidence_threshold=confidence_threshold,
        timeout=timeout,
        chat_fn=chat_fn,
    )

File: src/decider.rs
//! Decision engine. Mirrors `decider.py`.
//!
//! Two paths:
//!   - [`rule_decide`]: cheap, no model call. Returns `Some(Decision)` when
//!     the answer is unambiguous (Recommended marker, cursor-on-Yes,
//!     canonical Yes/No).
//!   - [`llm_decide`]: HTTP call to a local Ollama text model.
//!
//! [`decide`] is the top-level: rule path first, LLM as fallback.

use std::sync::OnceLock;

use regex::Regex;
use serde::Serialize;

use crate::parser::{Kind, Option, PromptFrame};

pub const ALLOWED_ACTIONS: &[&str] = &["1", "2", "3", "4", "5", "y", "n", "a", "enter", "none"];

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Decision {
    pub action: String,
    pub confidence: f32,
    pub reason: String,
    pub source: Source,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Source {
    Rule,
    Llm,
    Fallback,
}

impl Decision {
    pub fn rule(action: &str, confidence: f32, reason: impl Into<String>) -> Self {
        Self { action: action.into(), confidence, reason: reason.into(), source: Source::Rule }
    }
    pub fn fallback(action: &str, confidence: f32, reason: impl Into<String>) -> Self {
        Self { action: action.into(), confidence, reason: reason.into(), source: Source::Fallback }
    }
}

fn safe_first_label() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"(?i)^\s*(yes|allow|ok|continue|proceed)\b").unwrap())
}
fn unsafe_first_label() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        Regex::new(r"(?i)don'?t ask again|for the rest of|always allow|never ask").unwrap()
    })
}
fn canonical_yes_label() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| {
        Regex::new(r"(?i)^(yes|yes,?\s*proceed|yes,?\s*continue|allow|ok|continue|proceed)$")
            .unwrap()
    })
}
fn canonical_no_prefix() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"(?i)^(no|cancel|abort)\b").unwrap())
}

/// Returns the safest option to auto-pick, or None.
fn safe_default_option(frame: &PromptFrame) -> std::option::Option<&Option> {
    let cand = frame
        .options
        .iter()
        .find(|o| o.key == "1")
        .or_else(|| frame.options.iter().find(|o| o.key == "y"))?;
    if unsafe_first_label().is_match(&cand.label) {
        return None;
    }
    let cand_label = cand.label.trim().to_lowercase();
    let keys: std::collections::BTreeSet<&str> =
        frame.options.iter().map(|o| o.key.as_str()).collect();

    // (a) Inline (y/n) — synthesised by the parser with labels "yes"/"no".
    if frame.options.len() == 2
        && keys.contains("y")
        && keys.contains("n")
        && cand.key == "y"
        && cand_label == "yes"
    {
        return Some(cand);
    }

    // (b) Cursor on a Yes-style option.
    if cand.is_cursor && safe_first_label().is_match(&cand.label) {
        return Some(cand);
    }

    // (c) Exactly two options, canonical Yes / No-prefixed labels.
    if frame.options.len() == 2
        && cand.key == "1"
        && canonical_yes_label().is_match(&cand_label)
    {
        if let Some(other) = frame.options.iter().find(|o| o.key == "2") {
            if canonical_no_prefix().is_match(other.label.trim()) {
                return Some(cand);
            }
        }
    }
    None
}

/// Cheap, deterministic decisions. Returns `None` if the LLM is needed.
pub fn rule_decide(frame: &PromptFrame) -> std::option::Option<Decision> {
    if frame.kind != Kind::Choice {
        return Some(Decision::rule(
            "none",
            1.0,
            format!("kind={:?} → no auto-keypress", frame.kind),
        ));
    }
    if frame.destructive {
        return Some(Decision::rule(
            "none",
            1.0,
            "destructive keyword detected — handing to human",
        ));
    }
    if frame.options.is_empty() {
        return Some(Decision::rule("none", 1.0, "no options parsed"));
    }

    let recommended: Vec<&Option> = frame.options.iter().filter(|o| o.recommended).collect();
    if recommended.len() == 1 {
        let r = recommended[0];
        return Some(Decision::rule(
            &r.key,
            0.95,
            format!("option {:?} marked recommended", r.key),
        ));
    }

    if let Some(safe) = safe_default_option(frame) {
        return Some(Decision::rule(
            &safe.key,
            0.85,
            format!("safe single-shot Yes on key {:?}", safe.key),
        ));
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::parse_prompt;

    fn lines(s: &[&str]) -> Vec<String> {
        s.iter().map(|x| x.to_string()).collect()
    }

    #[test]
    fn recommended_marker_decides() {
        let f = parse_prompt(&lines(&[
            "Choose?",
            "  1. Yes",
            "  2. Yes, allow all (recommended)",
            "  3. No",
        ]));
        let d = rule_decide(&f).expect("rule should decide");
        assert_eq!(d.action, "2");
        assert!(d.confidence > 0.9);
    }

    #[test]
    fn cursor_on_yes_decides() {
        let f = parse_prompt(&lines(&[
            "Proceed?",
            "❯ 1. Yes",
            "  2. Yes, and don't ask again",
            "  3. No, with feedback",
        ]));
        let d = rule_decide(&f).expect("rule should decide");
        assert_eq!(d.action, "1");
    }

    #[test]
    fn yn_inline_decides_y() {
        let f = parse_prompt(&lines(&["Apply changes? (y/N)"]));
        let d = rule_decide(&f).expect("rule should decide");
        assert_eq!(d.action, "y");
    }

    #[test]
    fn destructive_returns_none() {
        let f = parse_prompt(&lines(&[
            "Force push to main?",
            "❯ 1. Yes",
            "  2. No",
        ]));
        let d = rule_decide(&f).expect("rule short-circuits destructive");
        assert_eq!(d.action, "none");
        assert!(d.reason.to_lowercase().contains("destructive"));
    }

    #[test]
    fn text_kind_returns_none() {
        let f = parse_prompt(&lines(&["What is your name?", "> "]));
        let d = rule_decide(&f).expect("text returns none");
        assert_eq!(d.action, "none");
    }

    #[test]
    fn ambiguous_falls_through_to_llm() {
        // Three options, no recommended marker, no cursor, label #1 has "don't ask again" → unsafe.
        let f = parse_prompt(&lines(&[
            "Pick one?",
            "  1. Yes, and don't ask again",
            "  2. Yes",
            "  3. No",
        ]));
        assert!(rule_decide(&f).is_none(), "must delegate to LLM");
    }

    #[test]
    fn two_choice_yes_no_decides() {
        let f = parse_prompt(&lines(&[
            "Continue?",
            "  1. Yes",
            "  2. No",
        ]));
        let d = rule_decide(&f).expect("canonical yes/no");
        assert_eq!(d.action, "1");
    }
}
